### PR TITLE
Material Scanner

### PR DESCRIPTION
**Pull request by eeSPee**
## About The Pull Request

Adds material scanner, a nifty little device for scanning material value of items. 
Also adds design; printable by cargo, science and engineering.
2 of these item spawn on all map; in science lab and QM office.

## Why It's Good For The Game

Encourages recycling.

## Changelog
:cl:
add: Add material scanner, a nifty little device for scanning material value of items. Printable design and some spawn on the map.
/:cl:

### CODE REVIEW

File: code/game/objects/items/devices/scanners.dm [hunk 1]
```diff
@@ -966,3 +966,36 @@ GENE SCANNER
...
+	lefthand_file = 'icons/mob/inhands/equipment/tools_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/equipment/tools_righthand.dmi'
+	w_class = WEIGHT_CLASS_SMALL
+	flags_1 = CONDUCT_1
+	throwforce = 0
+	throw_speed = 3
+	throw_range = 7
+	materials = list(/datum/material/iron=30, /datum/material/glass=20)
+
+/obj/item/material_scanner/attack(obj/O, mob/living/user)
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
/obj/item/material_scanner/attack(obj/item/I, mob/living/user)
```

File: code/game/objects/items/devices/scanners.dm [hunk 2]
```diff
@@ -966,3 +966,36 @@ GENE SCANNER
...
+	if(user.stat || user.eye_blind)
+		return
+	if(istype(O, /obj/item))
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
	if(istype(I))
```

File: code/game/objects/items/devices/scanners.dm [hunk 3]
```diff
@@ -966,3 +966,36 @@ GENE SCANNER
...
+		var/obj/item/I = O
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
```

**Comment by eeSPee** [comment]
Thanks

File: code/game/objects/items/devices/scanners.dm [hunk 4]
```diff
@@ -966,3 +966,35 @@ GENE SCANNER
...
+	materials = list(/datum/material/iron=30, /datum/material/glass=20)
+
+/obj/item/material_scanner/attack(obj/item/I, mob/living/user)
+	if(user.stat || user.eye_blind)
+		return
+	if(istype(I))
+		if (I.materials)
+			analyze_materials(I, user)
+			return	
+	to_chat(user, "<span class='notice'>[O] doesn't have any reusable materials.</span>")
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
	to_chat(user, "<span class='notice'>[I] doesn't have any reusable materials.</span>")
```

File: code/game/objects/items/devices/scanners.dm [hunk 5]
```diff
@@ -966,3 +966,35 @@ GENE SCANNER
...
+			return	
+	to_chat(user, "<span class='notice'>[I] doesn't have any reusable materials.</span>")
+
+/obj/item/material_scanner/proc/analyze_materials(obj/item/I, mob/living/user)
+	to_chat(user, "- Material Analysis of [I] -")
+	for(var/R in I.materials)
+		var/datum/material/M = R
+		var/amount =  I.materials[R]
+		to_chat(user, "* [M.name] ([amount])")
+	to_chat(user, "- Analysis complete! -")
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/scanners.dm]: I'd recommend using some spans so it's not just plain white text, maybe <span class='notice'>Or something along them lines</span>

Comment (eeSPee) [reply code/game/objects/items/devices/scanners.dm]: yea that's a good idea.

Comment (eeSPee) [reply code/game/objects/items/devices/scanners.dm]: can you make switch cases in bycode?

Comment (PowerfulBacon) [reply code/game/objects/items/devices/scanners.dm]: Switches work in byond

**Review by ike709 — CHANGES_REQUESTED** [review]
Two things:

1. Post a pic of what the output looks like on something with a variety of materials.
2. Push an empty commit so Travis runs again.

File: html/changelog.html [hunk 6]
```diff
@@ -848,6 +848,14 @@ <h3 class="author">super12pl updated:</h3>
 			<ul class="changes bgimages16">
 				<li class="rscadd">Added science sunglasses</li>
 			</ul>
+
+			<h2 class="date">18 August 2020</h2>
```
Comment (ike709) [inline html/changelog.html]: You shouldn't ever have diffs in this file.

**Comment by eeSPee** [comment]
This is with 3 mats
image

Not sure what to scan with more than 3 materials to so you can see it properly.

File: code/game/objects/items/devices/scanners.dm [hunk 7]
```diff
@@ -966,3 +967,65 @@ GENE SCANNER
...
+		else if (istype(M,/datum/material/bananium))
+			fcolor = "#ffff00"
+		else if (istype(M,/datum/material/titanium))
+			fcolor = "#dddddd"
+		output+= " - <font color='[fcolor]'>[M.name] ([amount])</font>"
+		integ++
+		if (integ>3)
+			to_chat(user, output + " - ")
+			integ = 1
+			output = ""
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/scanners.dm:1020]: What is this and why?

File: code/game/objects/items/devices/scanners.dm [hunk 8]
```diff
@@ -966,3 +967,65 @@ GENE SCANNER
...
+		if (I.materials && length(I.materials)>0)
+			analyze_materials(I, user)
+			return
+	to_chat(user, "<span class='notice'>[target] doesn't have any reusable materials.</span>")
+
+/obj/item/material_scanner/proc/analyze_materials(obj/item/I, mob/living/user)
+	to_chat(user, "- Material Analysis of [I] -")
+	var/integ = 1
+	var/output = ""
+	for(var/R in I.materials)
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm:1010]: This should definitely be an assoc list instead of an if/else if chain.

File: code/game/objects/items/devices/scanners.dm (hunk 7 above)
Comment (eeSPee) [reply code/game/objects/items/devices/scanners.dm:1020]: So it outputs:
-mat 1 - mat 2 - mat 3 -
-mat 4 - mat 5 - mat 6 -

instead of
-mat 1 - mat 2 - mat 3 - mat 4 - mat 5 - mat 6 -

File: code/game/objects/items/devices/scanners.dm (hunk 8 above)
Comment (eeSPee) [reply code/game/objects/items/devices/scanners.dm:1010]: Assoc list? I don't think I've done those before. Can you give an example?

**Comment by eeSPee** [comment]
Like this?

File: code/datums/components/crafting/tailoring.dm [hunk 9]
```diff
@@ -162,3 +162,12 @@
 				/obj/item/stack/sheet/glass = 1)
 	category = CAT_CLOTHING
 
+
+/datum/crafting_recipe/medsechud
```
Comment (ike709) [inline code/datums/components/crafting/tailoring.dm:166]: Unrelated change.

File: code/game/objects/items/devices/scanners.dm [hunk 10]
```diff
@@ -965,3 +966,58 @@ GENE SCANNER
...
+		var/datum/material/M = R
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm:1011]: Just do an `as()` typeless loop.

File: code/game/objects/items/devices/scanners.dm [hunk 11]
```diff
@@ -965,3 +966,58 @@ GENE SCANNER
...
+		var/amount =  I.materials[M]
+		var/fcolor = color_list[M]
+
```
Comment (ike709) [inline code/game/objects/items/devices/scanners.dm:1014]: You need sanity/a default color, in case the material isn't in the list.

File: html/changelog.html [hunk 12]
```diff
@@ -56,12 +56,6 @@
 		-->
 		<div class="commit sansserif">
 
-			<h2 class="date">24 November 2020</h2>
```
Comment (ike709) [inline html/changelog.html:59]: Don't touch this file.

File: html/changelogs/.all_changelog.yml [hunk 13]
```diff
@@ -28212,6 +28212,3 @@ DO NOT EDIT THIS FILE BY HAND!  AUTOMATICALLY GENERATED BY ss13_genchangelog.py.
   eeSPee:
   - balance: cult floor is atmosproof
   - balance: artificer magic missile no longer stuns cultists
-2020-11-24:
```
Comment (ike709) [inline html/changelogs/.all_changelog.yml:28215]: Don't touch this file.

File: html/changelogs/AutoChangeLog-pr-3002.yml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+author: jupyterkat
```
Comment (ike709) [inline html/changelogs/AutoChangeLog-pr-3002.yml:1]: Don't touch this file.
